### DISCUSSION TITLE
WT32-ETH01 board Ethernet library compatibility ? (LAN8720a chip)

### DISCUSSION THREAD

**Discussion by antoinec4** [eModbus/eModbus#338] [Q&A] [answered]
Hi,

I am working on a project with WT32-ETH01 boards in platformIO using the arduino framework ([datasheet](https://files.seeedstudio.com/products/102991455/WT32-ETH01_datasheet_V1.1-%20en.pdf)).
I have successfully integrated the eModbus library for modbus TCP communications over Wi-Fi (thanks to your great documentation).

Now, I would like to try modbus over ethernet since these boards have an on-board RJ45 port and ethernet chip (LAN8720A).
I have seen that it is recommended to use the Ethernet library by Max Gerhardt and I do have the lib folder "custom-Ethernet".
`#include "Ethernet.h"` does indeed include the right library, however I'm afraid it is incompatible with the LAN8720A since at first glance it looks made for Wiznet W5000 series chips.

Alternatively, I have successfully connected the WT32-ETH01 board to my LAN via Ethernet using the standard ETH.h lib, but I don't have communication via modbus.

Being rather new to using esp32, can someone more experienced confirm my thoughts or point me in the right direction?

Picture of the setup :
![Screenshot 2024-02-18 115301](https://github.com/eModbus/eModbus/assets/124214296/cb6ddba2-1a3d-497a-8f8b-6ae80be9f96e)

Thanks,

**Comment by Miq1** [comment]
I am afraid you will have to find out what may be missing from the standard ``ETH.h`` to prevent eModbus to successfully communicate. I do not have that chip variety in stock, so I can not test it myself. I remotely remember that I had to use Max' Ethernet lib due to some bug in the standard lib that was discussed between the maintainers of the Arduino core lib and the ESP IDF team to find the responsibility - without any chance to find a solution at that time.

**Reply by Miq1** [reply]
Addendum: out of curiosity I ordered one of these boards myself. The delivery was announced for mid-March, so it may take quite some time before I can put my hands on it. 
In the meantime feel free to keep me posted and ask whatever you need to know! 😉

**Reply by antoinec4** [reply]
I ordered mine on  [Digikey](https://www.digikey.ca/en/products/detail/seeed-technology-co-ltd/102991455/13572085) and the delivery was very fast, if that's an option for you!

After researching, I've just realized that there was another user who stumbled on the same dilemma as me :
[link](https://github.com/eModbus/eModbus/issues/209)

The modifications needed might be too ambitious for my skill set and I'm not limited to using this specific board, although the packaging was interesting. Ethernet connectivity is a requirement for my project.

What are some other boards and ethernet adapters that work and that you might recommend ?
Would using a standard esp32 dev board + a cheap W5500 SPI adapter board work ?

Thanks for the quick reply!

-Antoine

**Comment by Miq1** [comment]
> What are some other boards and ethernet adapters that work and that you might recommend ?
> Would using a standard esp32 dev board + a cheap W5500 SPI adapter board work ?

That is in fact the combination I am using and is working fine.

**Reply by antoinec4** [reply]
Great, I just ordered these cheap two adapters (same thing, different board):
[adapter 1](https://www.amazon.ca/-/fr/gp/product/B09ZSN4KYW/)
[adapter 2](https://www.amazon.ca/-/fr/gp/product/B0CKT2FWYG/)

I should receive them on Tuesday if all go well. I'll keep you posted if that works as intended.

In the mean time, if you'd like to test some code for the WT32 board, I could test it here, just reach out.

Have a nice day,

**Reply by Miq1** [reply]
👍🏻

**Reply by antoinec4** [reply]
@Miq1 Just to keep you updated, I managed to made it work quite well with a standard esp32 dev board and the bought adapters. Still no luck with the WT32 board though.

**Reply by Miq1** [reply]
Thanks for the update. I did not get my board yet, but will report my proceedings.

**Reply by Miq1** [reply]
Just a life sign: I got the board in the meantime but was not able to flash it (the FTDI adapter has both RX and TX LEDs constantly lighting and vanishes from the OS as soon as I will connect the WT32). I ordered another... 🙄

**Reply by antoinec4** [reply]
I also had problems with usb adapters, I figured if you use one that has a supply of 3.3V it work better.

**Comment by Miq1** [comment]
Got it to flash at least. There is some issue with the Ethernet server, ATM I only managed to get it into a boot loop when ``Ethernet.begin(mac);`` is called. Will dive into it tomorrow again.

**Reply by Miq1** [reply]
It looks like there is no plain ``EthernetServer`` and ``EthernetClient`` replacement. The [WT32_ETH01_WebServer by @khoih-prog](https://github.com/khoih-prog/WebServer_WT32_ETH01) has several additions to facilitate a HTTP server we do not need. I will dive into that nevertheless to see if I can extract just those parts we need here.

**Reply by antoinec4** [reply]
I also got into the boot loop problem with the Ethernet.begin method.
I don't expect you to solve my personal issue, especially if that requires an unreasonable amount of work, but it's nonetheless greatly appreciated.

On a positive note, I got the library working quite good with a regular esp32 dev board and the w5500 adapter.
Are there other adapters you did try ?

**Reply by Miq1** [reply]
No, there simply was no need yet. Many applications in fact are based on the bridge feature, so it seems, that was my initial application as well. And I took what was available for an Ethernet-RTU bridge then - thus the W5500. 😁

**Comment by Miq1** [comment]
It was easier than I had thought 😁 

``ETH.h`` quietly is supporting ``WiFiServer`` and ``WiFiClient`` objects, so all I had to do was to create a ``ModbusServerETH.h`` file like
```cpp
// =================================================================================================
// eModbus: Copyright 2024 by Michael Harwerth, Bert Melis and the contributors to eModbus
//               MIT license - see license.md for details
// =================================================================================================
#ifndef _MODBUS_SERVER_ETH_H
#define _MODBUS_SERVER_ETH_H
#include "options.h"
#if HAS_ETHERNET == 1
#include <ETH.h>
#include <SPI.h>

#undef SERVER_END
#define SERVER_END // NIL for Ethernet

#include "ModbusServerTCPtemp.h"
using ModbusServerEthernet = ModbusServerTCP<WiFiServer, WiFiClient>;
#endif

#endif
```
The application the will have to include that instead of the Ethernet variety:
```cpp
#include <Arduino.h>
#include <ModbusServerETH.h>

ModbusServerEthernet MSE;

ModbusMessage FC03(ModbusMessage request) {
  return ECHO_RESPONSE;
}

void setup() {
  Serial.begin(115200);
  Serial.print  (__DATE__);
  Serial.print  (" ");
  Serial.println(__TIME__);

  int result = ETH.begin(ETH_PHY_ADDR, ETH_PHY_POWER);
  ETH.setHostname("WT32-01");

  Serial.print("result: "); Serial.println(result);
  Serial.print("ETH MAC: "); Serial.println(ETH.macAddress());
  Serial.print("IPv4: "); Serial.println(ETH.localIP());
  Serial.print("fullDuplex: "); Serial.println(ETH.fullDuplex());
  Serial.print("linkSpeed: "); Serial.println(ETH.linkSpeed());

  MSE.registerWorker(1, READ_HOLD_REGISTER, FC03);
  MSE.start(502, 4, 2000);
}

void loop() {
}
```

I will add the ETH files for Client, Bridge and Server soon to the master branch.

**Comment by Miq1** [comment, marked as answer]
The board now is supported with the ``ModbusServerETH.h`` component.

**Reply by antoinec4** [reply]
Thank you, awesome
